feat: Adds SSL mutual auth support to intra-cluster requests (#5482)
confluentinc/ksql: docs/operate-and-deploy/installation/server-config/security.md

keywords: ksqldb, confguration, security, acl, ssl, sasl, keystore, truststore
---

ksqlDB supports several combinations of encryption and authentication on its

> client-facing and internal endpoints, and also supports many of the security

Reviewer: ```suggestion
client-facing and internal endpoints. ksqlDB also supports many of the security
```

    - [Securing headless deployments](#securing-headless-deployments)

  - [Securing communication with other services](#securing-communication-with-other-services)
    - [Configure ksqlDB for Confluent Control Center](#configure-ksqldb-for-confluent-control-center)
    - [Configure ksqlDB for Secured Confluent Schema Registry](#configure-ksqldb-for-secured-confluent-schema-registry)
    - [Configure ksqlDB for Secured Apache Kafka clusters](#configure-ksqldb-for-secured-apache-kafka-clusters)
      - [Configuring Kafka Encrypted Communication](#configuring-kafa-encrypted-communication)
      - [Configuring Kafka Authentication](#configuring-kafka-authentication)
      - [Configure Authorization of ksqlDB with Kafka ACLs](#configure-authorization-of-ksqldb-with-kafka-acls)
        - [Confluent Platform v5.0 (Apache Kafka v2.0) and above](#confluent-platform-v5.0-(apache-kafka-v2.0)-and-above)
        - [Confluent Platform versions below v5.0 (Apache Kafka < v2.0)](#confluent-platform-versions-below-v.5.0-(apache-kafka-<-v2.0))

Securing ksqlDB on premise
--------------------------

This section covers how to secure installations of ksqlDB outside of Confluent Cloud,

> e.g. on-premise installations or manual installations on other cloud platforms.

Reviewer: ```suggestion
like on-premises installations or manual installations on other cloud platforms.
```

    or TAR archives. For more information, see
    [On-Premises Deployments](https://docs.confluent.io/current/installation/installing_cp/index.html).

The section is split into:

[Securing ksqlDB installation](#securing-ksqldb-installation): Covers how to secure access to ksqlDB itself
and its own internal communication.

[Securing communication with other services](#securing-communication-with-other-services): Covers how to secure
communication with other services and access to the resources they hold.

Securing ksqlDB installation
----------------------------

ksqlDB supports two main deployment modes:

[Securing interactive deployments](#securing-interactive-deployments): Interactive deployments are those where
the ksqlDB servers accept client connections.

[Securing headless deployments](#securing-headless-deployments): headless
deployments are those where the ksqlDB servers do not accept client connections.
They read the SQL statements they should run from a file on-disk.

Securing interactive deployments
--------------------------------

> Securing the interactive ksqlDB installation involves securing the HTTP endpoints the ksqlDB server is listening on.

Reviewer: ```suggestion
Securing the interactive ksqlDB installation involves securing the HTTP endpoints that the ksqlDB server is listening on.
```

As well as accepting connections and requests from clients, a multi-node ksqlDB
cluster also requires inter-node communications. You can choose to configure
the external client and internal inter-node communication separately or over
a single listener:

[Securing single listener setup](#securing-single-listener-setup): Ideal for
single-node installations, or where the inter-node communication is over the
same network interfaces as client communication.

[Securing dual listener setup](#securing-dual-listener-setup): Useful where
inter-node communication is over a different network interfaces or requires
different authentication or encryption configuration.

Securing headless deployments
-----------------------------

Unlike interactive deployments, there are no exposed REST APIs, so security
is greatly simplified.

You may still have to [secure communication](#securing-communication-with-other-services) with other services.

> Securing single listener setup

Reviewer: ```suggestion
Securing single-listener setup
```

<ksql-install>bin/ksql --user fred --password letmein http://localhost:8088
```

> Configure ksqlDB for Internal Authentication

Author: > I've just had a look at what's already in this document and I think, rather than adding a brand new section, the details you're covering need to be integrated with what is already there.

I followed your instructions fairly closely, so I think we're good there.

@JimGalasyn Can you take a look at my docs changes again?  Also, How can I verify that the local anchor links work correctly?  None of my tools allow for checking this.

ssl.truststore.password=zzzz

listeners=http://0.0.0.0:8088
ksql.internal.listener=https://node-1.internal.example.com:8099

ksql.internal.ssl.client.authentication=REQUIRED

authentication.method=BASIC
authentication.roles=admin,developer,user,ksq-user
authentication.realm=KsqlServer-Props

```

For more detail on basic authentication, 
[see above](#configure-ksqldb-for-basic-http-authentication).

### Configuring HTTPS on the Internal and External Listeners

If you want to use HTTPS on `listeners` as well as use SSL mutual
auth for internal communication on `ksql.internal.listener`, you will likely
require two different key pairs, since your host's identity to clients may be
different from its internal identity.  In order to create such a key store,
refer [below](#setting-up-a-key-store-and-trust-store).

In such a configuration, you must specify which key pair is used for a given
listener by providing a key store alias.  For example, 
if set, `ksql.ssl.keystore.alias.internal` will be used to find the key store entry
with the given alias when setting up the internal listener.  Similarly,
`ksql.ssl.keystore.alias.external` is used for the client listener `listeners`.

> Below is an example configuration:

Reviewer 2: perfect!

```properties
# Contains the client certificate for this node under the alias 'client'.
# (e.g. key pair for external.example.com)
# Contains the internal certificate for this node under the alias 'internal_node1'.
# (e.g. key pair for node-1.internal.example.com)
ssl.keystore.location=/var/private/ssl/ksql.server.keystore.jks
ssl.keystore.password=xxxx
ssl.key.password=yyyy

# The aliases referenced above
ksql.ssl.keystore.alias.external=client
ksql.ssl.keystore.alias.internal=internal_node1

# Contains the certificates for nodes in the cluster.
# (e.g. certs for node-1.internal.example.com, node-2.internal.example.com)
ssl.truststore.location=/var/private/ssl/ksql.server.truststore.jks
ssl.truststore.password=zzzz

listeners=https://external.example.com:8088
ksql.internal.listener=https://node-1.internal.example.com:8099

# This enables mutual auth checking for the internal listener

> ksql.internal.ssl.client.authentication=REQUIRED

Reviewer 2: Maybe include `ksql.ssl.client.authentication=` whatever, just so users are aware?

<ksql-install>bin/ksql --user fred --password letmein http://localhost:8088
```

Configure ksqlDB for Internal Authentication

> ------------------------------------

Reviewer: ```suggestion
--------------------------------------------
```

> Since some system requests aren't done on behalf of a user (e.g. `/heartbeat`, `/lag`), 

Reviewer: ```suggestion
Since some system requests aren't done on behalf of a user, like `/heartbeat` and `/lag`, 
```

contain not only the certificate/key for the endpoints exposed with `listeners`, but also
the additional certificate/key for your internal listener set with
`ksql.internal.listener`.  Similarly, since your internal certificate is likely
self-signed, a trust store is required to contain certificates for nodes in your cluster.

```properties
# Contains both internal and external certificates for this node.
# e.g. cert and key for external.example.com and node-1.internal.example.com
ssl.keystore.location=/var/private/ssl/ksql.server.keystore.jks
ssl.keystore.password=xxxx
ssl.key.password=yyyy

# Contains internal certificates for all nodes in the cluster.
# e.g. cert for node-2.internal.example.com and node-3.internal.example.com
ssl.truststore.location=/var/private/ssl/ksql.server.truststore.jks
ssl.truststore.password=zzzz

listeners=https://external.example.com:8088
ksql.internal.listener=https://node-1.internal.example.com:8099
ksql.internal.ssl.client.authentication=REQUIRED
```

Not only does this configure system authentication to be secure and authenticated at the
connection level, but ksqlDB then also considers these requests to be authenticated.

> You can read generally about configuring `ksql.internal.listener` [here](index.md).

Reviewer: ```suggestion
For more information about configuring `ksql.internal.listener`, see [Configure ksqlDB for HTTPS](index.html#configure-ksqldb-for-https).
```

<ksql-install>bin/ksql --user fred --password letmein http://localhost:8088
```

> Securing dual listener setup

Reviewer: ```suggestion
Securing dual-listener setup
```

----------------------------

Using dual listeners for ksqlDB is appropriate when the client and
inter-node communication utilize different authentication and security

> configurations.  This is most likely the case when ksqlDB is deployed as an

Reviewer: ```suggestion
configurations. This is most likely the case when ksqlDB is deployed as an
```

The supported setups are SSL-mutual auth for the internal communication
combined with SSL encryption and authentication for the external client:

[Configuring internal for SSL-mutual authentication](#configuring-intenral-for-ssl-mutual-authentication):
Creates secure and authenticated connections for inter-node communication,
but leaves the external
client API unsecured. This is most appropriate when clients are trusted, but
the internal APIs are protected from use.

[Configuring internal for SSL-mutual authentication + external for SSL encryption](#configuring-internal-for-ssl-mutual-authentication-+-external-for-ssl-encryption):
Creates secure and authenticated connections for inter-node
communication and uses SSL for the external client API. This is most
likely to be pair with authentication below.

[Configuring internal for SSL-mutual authentication + external for HTTP-BASIC authentication](#configuring-internal-for-ssl-mutual-authentication-+-external-for-http-basic-authentication):
Creates secure and authenticated connections for inter-node
communication and uses basic authentication for the external client API.
This is most likely to be paired with SSL above.

> Configuring internal for SSL-mutual authentication

Reviewer: "internal" doesn't seem quite right here.

For more information about configuring `ksql.internal.listener`, see [Configuring Listeners of a ksqlDB Cluster](index.md#configuring-listeners-of-a-ksqldb-cluster).

Your key store must contain the key pair for your internal listener set with
`ksql.internal.listener`.  If your internal certificate is not signed by a recognized
public Certificate Authority, a trust store is required to contain certificates for
nodes in your cluster.  Below is an example configuration:

```properties
# Contains the internal key pair for this node.
# (e.g. key pair for node-1.internal.example.com)
ssl.keystore.location=/var/private/ssl/ksql.server.keystore.jks
ssl.keystore.password=xxxx
ssl.key.password=yyyy

# Contains the certificates for nodes in the cluster.
# (e.g. certs for node-1.internal.example.com, node-2.internal.example.com)
ssl.truststore.location=/var/private/ssl/ksql.server.truststore.jks
ssl.truststore.password=zzzz

listeners=http://0.0.0.0:8088
ksql.internal.listener=https://node-1.internal.example.com:8099

# This enables mutual auth checking for the internal listener
ksql.internal.ssl.client.authentication=REQUIRED
```

> Configuring internal for SSL-mutual authentication + external for HTTP-BASIC authentication

Reviewer: ```suggestion
Configuring internal for SSL-mutual authentication and external for HTTP-BASIC authentication
```

> --------------------------------------------------------------------------------

Reviewer: ```suggestion
---------------------------------------------------------------------------------------------
```

> Client facing basic HTTP authentication can be used alongside authentication for the

Reviewer: ```suggestion
Client-facing basic HTTP authentication can be used alongside authentication for the
```

internal listener. This ensures that neither the client or internal
APIs can be accessed by unauthorized users.

!!! note
	If you're using Basic authentication, we recommended that you
    [configure ksqlDB to use HTTPS for secure communication](#configuring-listner-for-ssl-encryption),
    because the Basic protocol passes credentials in plain text.

Below is an example configuration:

```properties
ssl.keystore.location=/var/private/ssl/ksql.server.keystore.jks
ssl.keystore.password=xxxx
ssl.key.password=yyyy

ssl.truststore.location=/var/private/ssl/ksql.server.truststore.jks
ssl.truststore.password=zzzz

listeners=http://0.0.0.0:8088
ksql.internal.listener=https://node-1.internal.example.com:8099

ksql.internal.ssl.client.authentication=REQUIRED

authentication.method=BASIC
authentication.roles=admin,developer,user,ksq-user
authentication.realm=KsqlServer-Props

```

For more detail on basic authentication,
[see above](#configure-ksqldb-for-basic-http-authentication).

> Configuring internal for SSL-mutual authentication + external for SSL encryption

Reviewer: ```suggestion
Configuring internal for SSL-mutual authentication and external for SSL encryption
```

> --------------------------------------------------------------------------------

Reviewer: ```suggestion
------------------------------------------------------------------------------------
```

If you want to use HTTPS on `listeners` as well as use SSL mutual
auth for internal communication on `ksql.internal.listener`, you will likely
require two different key pairs, since your host's identity to clients may be

> different from its internal identity.  In order to create such a key store,

Reviewer: ```suggestion
different from its internal identity. To create such a key store,
```

refer [below](#setting-up-a-key-store-and-trust-store).

In such a configuration, you must specify which key pair is used for a given

> listener by providing a key store alias.  For example,

Reviewer: ```suggestion
listener by providing a key store alias.  For example,
```
```suggestion
listener by providing a key store alias. For example,
```

if set, `ksql.ssl.keystore.alias.internal` will be used to find the key store entry

> with the given alias when setting up the internal listener.  Similarly,

Reviewer: ```suggestion
with the given alias when setting up the internal listener. Similarly,
```
